<a href="https://colab.research.google.com/github/sk27110/DL2/blob/HW-6-8/%D0%94%D0%976_8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Протестируем модели с классическим вниманием и вниманием из статьи на корпусе Flowers 102. Это корпус для fine-grained многоклассовой классификации. Посмотрим на обучение моделей при примерно равном числе параметров, замерим accuracy, время на эпоху и использование памяти. Реализацию MHLA возьмем свою, так как оригинальный код не запускается из-за проблем с библиотеками.

In [1]:
import torch

print("Torch version:", torch.__version__)
print("CUDA available:", torch.cuda.is_available())

if torch.cuda.is_available():
    print("GPU:", torch.cuda.get_device_name(0))
    print("CUDA version:", torch.version.cuda)


Torch version: 2.11.0+cu128
CUDA available: True
GPU: Tesla T4
CUDA version: 12.8


In [2]:
import os
import time
import math
import random
import copy
from dataclasses import dataclass

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from tqdm.auto import tqdm

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, Subset

import torchvision
import torchvision.transforms as T
from torchvision.datasets import Flowers102

from torch.cuda.amp import autocast, GradScaler


In [3]:
def set_seed(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

    torch.backends.cudnn.benchmark = True
    torch.backends.cudnn.deterministic = False

set_seed(42)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
device


device(type='cuda')

In [4]:
@dataclass
class Config:
    data_root: str = "./data"

    image_size: int = 128
    patch_size: int = 4
    num_classes: int = 102

    batch_size: int = 64
    num_workers: int = 0

    epochs: int = 60
    lr: float = 4e-4
    weight_decay: float = 0.03

    embed_dim: int = 128
    depth: int = 3
    num_heads: int = 4
    mlp_ratio: float = 4.0
    dropout: float = 0.1

    use_amp: bool = False


cfg = Config()

print(cfg)
print("Number of image tokens:", (cfg.image_size // cfg.patch_size) ** 2)


Config(data_root='./data', image_size=128, patch_size=4, num_classes=102, batch_size=64, num_workers=0, epochs=60, lr=0.0004, weight_decay=0.03, embed_dim=128, depth=3, num_heads=4, mlp_ratio=4.0, dropout=0.1, use_amp=False)
Number of image tokens: 1024


In [5]:
train_transform = T.Compose([
    T.Resize((cfg.image_size, cfg.image_size)),
    T.RandomHorizontalFlip(p=0.5),
    T.RandomRotation(15),
    T.ColorJitter(brightness=0.2, contrast=0.2, saturation=0.2, hue=0.05),
    T.ToTensor(),
    T.Normalize(mean=(0.485, 0.456, 0.406),
                std=(0.229, 0.224, 0.225)),
])

eval_transform = T.Compose([
    T.Resize((cfg.image_size, cfg.image_size)),
    T.ToTensor(),
    T.Normalize(mean=(0.485, 0.456, 0.406),
                std=(0.229, 0.224, 0.225)),
])

train_dataset = Flowers102(
    root=cfg.data_root,
    split="train",
    download=True,
    transform=train_transform,
)

val_dataset = Flowers102(
    root=cfg.data_root,
    split="val",
    download=True,
    transform=eval_transform,
)

test_dataset = Flowers102(
    root=cfg.data_root,
    split="test",
    download=True,
    transform=eval_transform,
)

print("Train:", len(train_dataset))
print("Val:", len(val_dataset))
print("Test:", len(test_dataset))


100%|██████████| 345M/345M [00:12<00:00, 26.5MB/s]
100%|██████████| 502/502 [00:00<00:00, 877kB/s]
100%|██████████| 15.0k/15.0k [00:00<00:00, 43.7MB/s]


Train: 1020
Val: 1020
Test: 6149


In [6]:
train_loader = DataLoader(
    train_dataset,
    batch_size=cfg.batch_size,
    shuffle=True,
    num_workers=0,
    pin_memory=(device.type == "cuda"),
    drop_last=False,
)

val_loader = DataLoader(
    val_dataset,
    batch_size=cfg.batch_size,
    shuffle=False,
    num_workers=0,
    pin_memory=(device.type == "cuda"),
    drop_last=False,
)

test_loader = DataLoader(
    test_dataset,
    batch_size=cfg.batch_size,
    shuffle=False,
    num_workers=0,
    pin_memory=(device.type == "cuda"),
    drop_last=False,
)


In [16]:
class PatchEmbed(nn.Module):
    def __init__(self, image_size=128, patch_size=8, in_chans=3, embed_dim=192):
        super().__init__()

        assert image_size % patch_size == 0, "image_size must be divisible by patch_size"

        self.image_size = image_size
        self.patch_size = patch_size
        self.grid_size = image_size // patch_size
        self.num_patches = self.grid_size * self.grid_size

        self.proj = nn.Conv2d(
            in_chans,
            embed_dim,
            kernel_size=patch_size,
            stride=patch_size,
        )

    def forward(self, x):
        x = self.proj(x)
        x = x.flatten(2).transpose(1, 2)
        return x


In [17]:
class StandardAttention(nn.Module):
    def __init__(self, dim, num_heads=3, attn_drop=0.0, proj_drop=0.0):
        super().__init__()

        assert dim % num_heads == 0

        self.dim = dim
        self.num_heads = num_heads
        self.head_dim = dim // num_heads
        self.scale = self.head_dim ** -0.5

        self.qkv = nn.Linear(dim, dim * 3, bias=True)
        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def forward(self, x):
        B, N, C = x.shape

        qkv = self.qkv(x)
        qkv = qkv.reshape(B, N, 3, self.num_heads, self.head_dim)
        qkv = qkv.permute(2, 0, 3, 1, 4)

        q, k, v = qkv[0], qkv[1], qkv[2]
        attn = (q @ k.transpose(-2, -1)) * self.scale
        attn = attn.softmax(dim=-1)
        attn = self.attn_drop(attn)

        out = attn @ v
        out = out.transpose(1, 2).reshape(B, N, C)

        out = self.proj(out)
        out = self.proj_drop(out)

        return out


In [18]:
class SequenceLabZeroSAttention(nn.Module):
    def __init__(
        self,
        dim,
        num_heads=4,
        dropout=0.0,
        block_size=1025,
        bias=True,
    ):
        super().__init__()

        from sequencelab.build import build_attention
        from sequencelab.config import ZeroSConfig

        cfg = ZeroSConfig(
            n_embd=dim,
            n_head=num_heads,
            dropout=dropout,
            bias=bias,
            block_size=block_size,
            is_causal=False,
            init_params=False,
            init_n_layers=1,
            use_norm=True,
            use_associative=True,
        )

        self.attn = build_attention(cfg)

    def forward(self, x):
        return self.attn(x)


In [22]:
class SequenceLabFEMAttention(nn.Module):
    def __init__(
        self,
        dim,
        num_heads=4,
        dropout=0.0,
        prior_type="softmax",
        bias=True,
    ):
        super().__init__()

        from sequencelab.build import build_attention
        from sequencelab.config import FEMConfig

        cfg = FEMConfig(
            n_embd=dim,
            n_head=num_heads,
            prior_type=prior_type,
            dropout=dropout,
            causal=False,
            bias=bias,

            fem_ratio=0.5,
            p_t_to_fem_ratio=2.0,

            use_temperature=True,
            use_lse=True,
            use_outer_gate=True,

            use_rope=False,
            use_conv=False,
            conv_hidden=64,
            conv_norm_first=True,
            conv_bidirectional=False,
        )

        self.attn = build_attention(cfg)

    def forward(self, x):
        return self.attn(x)


In [38]:
class ELSAAttention(nn.Module):
    def __init__(
        self,
        dim,
        num_heads=4,
        attn_drop=0.0,
        proj_drop=0.0,
        backend="auto",
    ):
        super().__init__()

        from stable.elsa import ElsaAttention

        self.attn = ElsaAttention(
            dim=dim,
            num_heads=num_heads,
            attn_drop=attn_drop,
            proj_drop=proj_drop,
            backend=backend,
        )

    def forward(self, x):
        return self.attn(x)


In [19]:
class MHLALinearAttention(nn.Module):
    def __init__(
        self,
        dim,
        num_heads=3,
        attn_drop=0.0,
        proj_drop=0.0,
        eps=1e-6,
    ):
        super().__init__()

        assert dim % num_heads == 0

        self.dim = dim
        self.num_heads = num_heads
        self.head_dim = dim // num_heads
        self.eps = eps

        self.qkv = nn.Linear(dim, dim * 3, bias=True)

        self.head_gate = nn.Linear(dim, num_heads, bias=True)

        self.local_conv = nn.Conv2d(
            dim,
            dim,
            kernel_size=3,
            padding=1,
            groups=dim,
            bias=True,
        )

        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def feature_map(self, x):
        return F.elu(x) + 1.0

    def local_positional_encoding(self, x):
        B, N, C = x.shape

        cls_token = x[:, :1, :]
        patch_tokens = x[:, 1:, :]

        num_patches = patch_tokens.shape[1]
        grid_size = int(math.sqrt(num_patches))

        if grid_size * grid_size != num_patches:
            return torch.zeros_like(x)

        feat = patch_tokens.transpose(1, 2).reshape(B, C, grid_size, grid_size)
        feat = self.local_conv(feat)
        feat = feat.flatten(2).transpose(1, 2)

        zero_cls = torch.zeros_like(cls_token)
        return torch.cat([zero_cls, feat], dim=1)

    def forward(self, x):
        B, N, C = x.shape

        local = self.local_positional_encoding(x)

        qkv = self.qkv(x)
        qkv = qkv.reshape(B, N, 3, self.num_heads, self.head_dim)
        qkv = qkv.permute(2, 0, 3, 1, 4)

        q, k, v = qkv[0], qkv[1], qkv[2]  # [B, H, N, D]

        q = self.feature_map(q)
        k = self.feature_map(k)

        v = self.attn_drop(v)

        kv = torch.einsum("bhnd,bhne->bhde", k, v)

        k_sum = k.sum(dim=2)
        z = 1.0 / (torch.einsum("bhnd,bhd->bhn", q, k_sum) + self.eps)

        out = torch.einsum("bhnd,bhde,bhn->bhne", q, kv, z)

        gates = torch.softmax(self.head_gate(x), dim=-1)

        out = out.permute(0, 2, 1, 3)
        out = out * gates.unsqueeze(-1) * self.num_heads
        out = out.reshape(B, N, C)

        out = out + local

        out = self.proj(out)
        out = self.proj_drop(out)

        return out


In [40]:
class MLP(nn.Module):
    def __init__(self, dim, hidden_dim, dropout=0.0):
        super().__init__()

        self.net = nn.Sequential(
            nn.Linear(dim, hidden_dim),
            nn.GELU(),
            nn.Dropout(dropout),
            nn.Linear(hidden_dim, dim),
            nn.Dropout(dropout),
        )

    def forward(self, x):
        return self.net(x)


class TransformerBlock(nn.Module):
    def __init__(
        self,
        dim,
        num_heads,
        mlp_ratio=4.0,
        dropout=0.1,
        attention_type="standard",
    ):
        super().__init__()

        self.norm1 = nn.LayerNorm(dim)

        if attention_type == "standard":
            self.attn = StandardAttention(
                dim=dim,
                num_heads=num_heads,
                attn_drop=dropout,
                proj_drop=dropout,
            )

        elif attention_type == "mhla":
            self.attn = MHLALinearAttention(
                dim=dim,
                num_heads=num_heads,
                attn_drop=dropout,
                proj_drop=dropout,
            )

        elif attention_type == "sequencelab_zeros":
            self.attn = SequenceLabZeroSAttention(
                dim=dim,
                num_heads=num_heads,
                dropout=dropout,
                block_size=2048,
                bias=True,
            )

        elif attention_type == "sequencelab_fem_softmax":
            self.attn = SequenceLabFEMAttention(
                dim=dim,
                num_heads=num_heads,
                dropout=dropout,
                prior_type="softmax",
                bias=True,
            )

        elif attention_type == "sequencelab_fem_linear":
            self.attn = SequenceLabFEMAttention(
                dim=dim,
                num_heads=num_heads,
                dropout=dropout,
                prior_type="linear",
                bias=True,
            )
        elif attention_type == "elsa_auto":
            self.attn = ELSAAttention(
                dim=dim,
                num_heads=num_heads,
                attn_drop=dropout,
                proj_drop=dropout,
                backend="auto",
            )

        elif attention_type == "elsa_triton":
            self.attn = ELSAAttention(
                dim=dim,
                num_heads=num_heads,
                attn_drop=dropout,
                proj_drop=dropout,
                backend="triton",
            )

        elif attention_type == "elsa_pytorch":
            self.attn = ELSAAttention(
                dim=dim,
                num_heads=num_heads,
                attn_drop=dropout,
                proj_drop=dropout,
                backend="pytorch",
            )

        elif attention_type == "elsa_triton_fp32_train":
            self.attn = ELSAAttention(
                dim=dim,
                num_heads=num_heads,
                attn_drop=dropout,
                proj_drop=dropout,
                backend="triton_fp32_train",
            )


        else:
            raise ValueError(f"Unknown attention_type: {attention_type}")

        self.norm2 = nn.LayerNorm(dim)

        hidden_dim = int(dim * mlp_ratio)
        self.mlp = MLP(
            dim=dim,
            hidden_dim=hidden_dim,
            dropout=dropout,
        )

    def forward(self, x):
        x = x + self.attn(self.norm1(x))
        x = x + self.mlp(self.norm2(x))
        return x


In [24]:
class TinyViT(nn.Module):
    def __init__(
        self,
        image_size=128,
        patch_size=8,
        in_chans=3,
        num_classes=102,
        embed_dim=192,
        depth=4,
        num_heads=3,
        mlp_ratio=4.0,
        dropout=0.1,
        attention_type="standard",
    ):
        super().__init__()

        self.patch_embed = PatchEmbed(
            image_size=image_size,
            patch_size=patch_size,
            in_chans=in_chans,
            embed_dim=embed_dim,
        )

        num_patches = self.patch_embed.num_patches

        self.cls_token = nn.Parameter(torch.zeros(1, 1, embed_dim))
        self.pos_embed = nn.Parameter(torch.zeros(1, num_patches + 1, embed_dim))
        self.pos_drop = nn.Dropout(dropout)

        self.blocks = nn.ModuleList([
            TransformerBlock(
                dim=embed_dim,
                num_heads=num_heads,
                mlp_ratio=mlp_ratio,
                dropout=dropout,
                attention_type=attention_type,
            )
            for _ in range(depth)
        ])

        self.norm = nn.LayerNorm(embed_dim)
        self.head = nn.Linear(embed_dim, num_classes)

        self.apply(self._init_weights)

    def _init_weights(self, m):
        if isinstance(m, nn.Linear):
            nn.init.trunc_normal_(m.weight, std=0.02)
            if m.bias is not None:
                nn.init.zeros_(m.bias)

        elif isinstance(m, nn.Conv2d):
            nn.init.kaiming_normal_(m.weight, mode="fan_out")
            if m.bias is not None:
                nn.init.zeros_(m.bias)

        elif isinstance(m, nn.LayerNorm):
            nn.init.ones_(m.weight)
            nn.init.zeros_(m.bias)

        nn.init.trunc_normal_(self.cls_token, std=0.02)
        nn.init.trunc_normal_(self.pos_embed, std=0.02)

    def forward(self, x):
        B = x.size(0)

        x = self.patch_embed(x)

        cls_tokens = self.cls_token.expand(B, -1, -1)
        x = torch.cat([cls_tokens, x], dim=1)

        x = x + self.pos_embed
        x = self.pos_drop(x)

        for block in self.blocks:
            x = block(x)

        x = self.norm(x)

        cls = x[:, 0]
        logits = self.head(cls)

        return logits


In [43]:
def count_params(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

standard_model = TinyViT(
    image_size=cfg.image_size,
    patch_size=cfg.patch_size,
    num_classes=cfg.num_classes,
    embed_dim=cfg.embed_dim,
    depth=cfg.depth,
    num_heads=cfg.num_heads,
    mlp_ratio=cfg.mlp_ratio,
    dropout=cfg.dropout,
    attention_type="standard",
)

mhla_model = TinyViT(
    image_size=cfg.image_size,
    patch_size=cfg.patch_size,
    num_classes=cfg.num_classes,
    embed_dim=cfg.embed_dim,
    depth=cfg.depth,
    num_heads=cfg.num_heads,
    mlp_ratio=cfg.mlp_ratio,
    dropout=cfg.dropout,
    attention_type="mhla",
)

seq_model = TinyViT(
    image_size=cfg.image_size,
    patch_size=cfg.patch_size,
    num_classes=cfg.num_classes,
    embed_dim=cfg.embed_dim,
    depth=cfg.depth,
    num_heads=cfg.num_heads,
    mlp_ratio=cfg.mlp_ratio,
    dropout=cfg.dropout,
    attention_type="sequencelab_fem_softmax",
)

elsa_model = TinyViT(
    image_size=cfg.image_size,
    patch_size=cfg.patch_size,
    num_classes=cfg.num_classes,
    embed_dim=cfg.embed_dim,
    depth=cfg.depth,
    num_heads=cfg.num_heads,
    mlp_ratio=cfg.mlp_ratio,
    dropout=cfg.dropout,
    attention_type="elsa_auto",
)

print("ELSA params:", count_params(elsa_model))



print("Standard params:", count_params(standard_model))
print("MHLA params:", count_params(mhla_model))
print("SequenceLab FEM params:", count_params(seq_model))


ELSA params: 744678
Standard params: 745830
MHLA params: 751218
SequenceLab FEM params: 696678


In [29]:
def accuracy_top1(logits, targets):
    preds = logits.argmax(dim=1)
    correct = (preds == targets).sum().item()
    total = targets.size(0)
    return correct, total


def train_one_epoch(
    model,
    loader,
    optimizer,
    criterion,
    scaler,
    device,
    use_amp=True,
):
    model.train()

    running_loss = 0.0
    running_correct = 0
    running_total = 0

    start_time = time.time()

    pbar = tqdm(loader, desc="train", leave=False)

    for images, targets in pbar:
        images = images.to(device, non_blocking=True)
        targets = targets.to(device, non_blocking=True)

        optimizer.zero_grad(set_to_none=True)

        with autocast(enabled=(use_amp and device.type == "cuda")):
            logits = model(images)
            loss = criterion(logits, targets)

        scaler.scale(loss).backward()
        scaler.step(optimizer)
        scaler.update()

        batch_size = targets.size(0)

        running_loss += loss.item() * batch_size
        correct, total = accuracy_top1(logits.detach(), targets)
        running_correct += correct
        running_total += total

        pbar.set_postfix({
            "loss": running_loss / running_total,
            "acc": running_correct / running_total,
        })

    epoch_time = time.time() - start_time

    return {
        "loss": running_loss / running_total,
        "acc": running_correct / running_total,
        "time": epoch_time,
    }


@torch.no_grad()
def evaluate(
    model,
    loader,
    criterion,
    device,
    use_amp=True,
    desc="eval",
):
    model.eval()

    running_loss = 0.0
    running_correct = 0
    running_total = 0

    start_time = time.time()

    pbar = tqdm(loader, desc=desc, leave=False)

    for images, targets in pbar:
        images = images.to(device, non_blocking=True)
        targets = targets.to(device, non_blocking=True)

        with autocast(enabled=(use_amp and device.type == "cuda")):
            logits = model(images)
            loss = criterion(logits, targets)

        batch_size = targets.size(0)

        running_loss += loss.item() * batch_size
        correct, total = accuracy_top1(logits, targets)
        running_correct += correct
        running_total += total

        pbar.set_postfix({
            "loss": running_loss / running_total,
            "acc": running_correct / running_total,
        })

    eval_time = time.time() - start_time

    return {
        "loss": running_loss / running_total,
        "acc": running_correct / running_total,
        "time": eval_time,
    }


In [42]:
def run_experiment(attention_type, cfg):
    allowed_attention_types = [
        "standard",
        "mhla",
        "sequencelab_zeros",
        "sequencelab_fem_softmax",
        "sequencelab_fem_linear",
        "elsa_auto",
        "elsa_triton",
        "elsa_pytorch",
        "elsa_triton_fp32_train",
    ]



    assert attention_type in allowed_attention_types


    set_seed(42)

    if torch.cuda.is_available():
        torch.cuda.empty_cache()
        torch.cuda.reset_peak_memory_stats()

    model = TinyViT(
        image_size=cfg.image_size,
        patch_size=cfg.patch_size,
        num_classes=cfg.num_classes,
        embed_dim=cfg.embed_dim,
        depth=cfg.depth,
        num_heads=cfg.num_heads,
        mlp_ratio=cfg.mlp_ratio,
        dropout=cfg.dropout,
        attention_type=attention_type,
    ).to(device)

    criterion = nn.CrossEntropyLoss(label_smoothing=0.1)

    optimizer = torch.optim.AdamW(
        model.parameters(),
        lr=cfg.lr,
        weight_decay=cfg.weight_decay,
    )

    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
        optimizer,
        T_max=cfg.epochs,
    )

    scaler = GradScaler(enabled=(cfg.use_amp and device.type == "cuda"))

    history = []

    best_val_acc = -1.0
    best_state = None

    print(f"\n===== Experiment: {attention_type} =====")
    print("Params:", count_params(model))

    total_start = time.time()

    for epoch in range(1, cfg.epochs + 1):
        print(f"\nEpoch {epoch}/{cfg.epochs}")

        train_metrics = train_one_epoch(
            model=model,
            loader=train_loader,
            optimizer=optimizer,
            criterion=criterion,
            scaler=scaler,
            device=device,
            use_amp=cfg.use_amp,
        )

        val_metrics = evaluate(
            model=model,
            loader=val_loader,
            criterion=criterion,
            device=device,
            use_amp=cfg.use_amp,
            desc="val",
        )

        scheduler.step()

        lr_now = optimizer.param_groups[0]["lr"]

        row = {
            "attention": attention_type,
            "epoch": epoch,
            "lr": lr_now,
            "train_loss": train_metrics["loss"],
            "train_acc": train_metrics["acc"],
            "train_time": train_metrics["time"],
            "val_loss": val_metrics["loss"],
            "val_acc": val_metrics["acc"],
            "val_time": val_metrics["time"],
        }

        history.append(row)

        print(
            f"train_loss={row['train_loss']:.4f} "
            f"train_acc={row['train_acc']:.4f} | "
            f"val_loss={row['val_loss']:.4f} "
            f"val_acc={row['val_acc']:.4f} | "
            f"lr={lr_now:.2e}"
        )

        if val_metrics["acc"] > best_val_acc:
            best_val_acc = val_metrics["acc"]
            best_state = copy.deepcopy(model.state_dict())

    total_time = time.time() - total_start

    if best_state is not None:
        model.load_state_dict(best_state)

    test_metrics = evaluate(
        model=model,
        loader=test_loader,
        criterion=criterion,
        device=device,
        use_amp=cfg.use_amp,
        desc="test",
    )

    if torch.cuda.is_available():
        peak_memory_mb = torch.cuda.max_memory_allocated() / 1024**2
    else:
        peak_memory_mb = 0.0

    summary = {
        "attention": attention_type,
        "params": count_params(model),
        "best_val_acc": best_val_acc,
        "test_loss": test_metrics["loss"],
        "test_acc": test_metrics["acc"],
        "total_train_time_sec": total_time,
        "avg_epoch_train_time_sec": np.mean([h["train_time"] for h in history]),
        "peak_memory_mb": peak_memory_mb,
    }

    return model, pd.DataFrame(history), summary


Проверим сперва модели на 30 эпохах.

In [80]:
mhla_model, mhla_history, mhla_summary = run_experiment(
    attention_type="mhla",
    cfg=cfg,
)

mhla_summary



===== Experiment: mhla =====
Params: 751218

Epoch 1/30


/tmp/ipykernel_7000/1961669753.py:35: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = GradScaler(enabled=(cfg.use_amp and device.type == "cuda"))


train:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_7000/1366650589.py:33: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


val:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_7000/1366650589.py:85: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


train_loss=4.6014 train_acc=0.0176 | val_loss=4.4626 val_acc=0.0382 | lr=3.99e-04

Epoch 2/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.4032 train_acc=0.0402 | val_loss=4.3382 val_acc=0.0500 | lr=3.96e-04

Epoch 3/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.2966 train_acc=0.0471 | val_loss=4.2683 val_acc=0.0686 | lr=3.90e-04

Epoch 4/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.2048 train_acc=0.0627 | val_loss=4.1949 val_acc=0.0716 | lr=3.83e-04

Epoch 5/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.1379 train_acc=0.0725 | val_loss=4.1404 val_acc=0.0843 | lr=3.73e-04

Epoch 6/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0710 train_acc=0.0951 | val_loss=4.0782 val_acc=0.0941 | lr=3.62e-04

Epoch 7/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0167 train_acc=0.0961 | val_loss=4.0288 val_acc=0.1108 | lr=3.49e-04

Epoch 8/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.9610 train_acc=0.1078 | val_loss=3.9748 val_acc=0.1255 | lr=3.34e-04

Epoch 9/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.9168 train_acc=0.1216 | val_loss=3.9310 val_acc=0.1206 | lr=3.18e-04

Epoch 10/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8628 train_acc=0.1206 | val_loss=3.8796 val_acc=0.1186 | lr=3.00e-04

Epoch 11/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7979 train_acc=0.1431 | val_loss=3.8558 val_acc=0.1314 | lr=2.81e-04

Epoch 12/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7712 train_acc=0.1431 | val_loss=3.8110 val_acc=0.1431 | lr=2.62e-04

Epoch 13/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7087 train_acc=0.1471 | val_loss=3.7875 val_acc=0.1461 | lr=2.42e-04

Epoch 14/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6860 train_acc=0.1647 | val_loss=3.7264 val_acc=0.1618 | lr=2.21e-04

Epoch 15/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6237 train_acc=0.1824 | val_loss=3.7123 val_acc=0.1725 | lr=2.00e-04

Epoch 16/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6070 train_acc=0.1794 | val_loss=3.6843 val_acc=0.1627 | lr=1.79e-04

Epoch 17/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5655 train_acc=0.2088 | val_loss=3.6705 val_acc=0.1667 | lr=1.58e-04

Epoch 18/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5551 train_acc=0.2020 | val_loss=3.6342 val_acc=0.1882 | lr=1.38e-04

Epoch 19/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5315 train_acc=0.2059 | val_loss=3.6246 val_acc=0.1922 | lr=1.19e-04

Epoch 20/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4962 train_acc=0.2167 | val_loss=3.6146 val_acc=0.1882 | lr=1.00e-04

Epoch 21/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4807 train_acc=0.2147 | val_loss=3.5970 val_acc=0.1912 | lr=8.24e-05

Epoch 22/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4592 train_acc=0.2373 | val_loss=3.5900 val_acc=0.1902 | lr=6.62e-05

Epoch 23/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4447 train_acc=0.2265 | val_loss=3.5688 val_acc=0.2020 | lr=5.14e-05

Epoch 24/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4157 train_acc=0.2461 | val_loss=3.5663 val_acc=0.1941 | lr=3.82e-05

Epoch 25/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4102 train_acc=0.2373 | val_loss=3.5569 val_acc=0.2088 | lr=2.68e-05

Epoch 26/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4025 train_acc=0.2569 | val_loss=3.5620 val_acc=0.1990 | lr=1.73e-05

Epoch 27/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3957 train_acc=0.2539 | val_loss=3.5564 val_acc=0.2029 | lr=9.79e-06

Epoch 28/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3959 train_acc=0.2471 | val_loss=3.5518 val_acc=0.2069 | lr=4.37e-06

Epoch 29/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3804 train_acc=0.2549 | val_loss=3.5515 val_acc=0.2039 | lr=1.10e-06

Epoch 30/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3945 train_acc=0.2539 | val_loss=3.5512 val_acc=0.2049 | lr=0.00e+00


test:   0%|          | 0/97 [00:00<?, ?it/s]

{'attention': 'mhla',
 'params': 751218,
 'best_val_acc': 0.2088235294117647,
 'test_loss': 3.691866265562302,
 'test_acc': 0.16815742397137745,
 'total_train_time_sec': 477.25461506843567,
 'avg_epoch_train_time_sec': np.float64(10.423238571484884),
 'peak_memory_mb': 2610.66650390625}

In [81]:
standard_model, standard_history, standard_summary = run_experiment(
    attention_type="standard",
    cfg=cfg,
)

standard_summary



===== Experiment: standard =====
Params: 745830

Epoch 1/30


/tmp/ipykernel_7000/1961669753.py:35: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = GradScaler(enabled=(cfg.use_amp and device.type == "cuda"))


train:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_7000/1366650589.py:33: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


val:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_7000/1366650589.py:85: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


train_loss=4.6036 train_acc=0.0137 | val_loss=4.4629 val_acc=0.0431 | lr=3.99e-04

Epoch 2/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.4088 train_acc=0.0382 | val_loss=4.3276 val_acc=0.0598 | lr=3.96e-04

Epoch 3/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.2786 train_acc=0.0569 | val_loss=4.2338 val_acc=0.0843 | lr=3.90e-04

Epoch 4/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.1824 train_acc=0.0696 | val_loss=4.1331 val_acc=0.0902 | lr=3.83e-04

Epoch 5/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0929 train_acc=0.0922 | val_loss=4.0624 val_acc=0.1147 | lr=3.73e-04

Epoch 6/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0270 train_acc=0.1147 | val_loss=3.9872 val_acc=0.1275 | lr=3.62e-04

Epoch 7/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.9500 train_acc=0.1353 | val_loss=3.9223 val_acc=0.1392 | lr=3.49e-04

Epoch 8/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8846 train_acc=0.1275 | val_loss=3.8707 val_acc=0.1608 | lr=3.34e-04

Epoch 9/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8075 train_acc=0.1676 | val_loss=3.7887 val_acc=0.1794 | lr=3.18e-04

Epoch 10/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7459 train_acc=0.1833 | val_loss=3.7439 val_acc=0.1863 | lr=3.00e-04

Epoch 11/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6751 train_acc=0.2029 | val_loss=3.7079 val_acc=0.1873 | lr=2.81e-04

Epoch 12/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6471 train_acc=0.1902 | val_loss=3.6530 val_acc=0.2010 | lr=2.62e-04

Epoch 13/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5890 train_acc=0.1961 | val_loss=3.6201 val_acc=0.1922 | lr=2.42e-04

Epoch 14/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5468 train_acc=0.2186 | val_loss=3.6048 val_acc=0.2049 | lr=2.21e-04

Epoch 15/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5119 train_acc=0.2186 | val_loss=3.5569 val_acc=0.2245 | lr=2.00e-04

Epoch 16/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4821 train_acc=0.2088 | val_loss=3.5617 val_acc=0.2127 | lr=1.79e-04

Epoch 17/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4438 train_acc=0.2333 | val_loss=3.5089 val_acc=0.2186 | lr=1.58e-04

Epoch 18/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4020 train_acc=0.2520 | val_loss=3.5053 val_acc=0.2314 | lr=1.38e-04

Epoch 19/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3819 train_acc=0.2549 | val_loss=3.4696 val_acc=0.2392 | lr=1.19e-04

Epoch 20/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3596 train_acc=0.2539 | val_loss=3.4501 val_acc=0.2412 | lr=1.00e-04

Epoch 21/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3153 train_acc=0.2735 | val_loss=3.4441 val_acc=0.2451 | lr=8.24e-05

Epoch 22/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2976 train_acc=0.2676 | val_loss=3.4282 val_acc=0.2490 | lr=6.62e-05

Epoch 23/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2902 train_acc=0.2931 | val_loss=3.4169 val_acc=0.2529 | lr=5.14e-05

Epoch 24/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2742 train_acc=0.2863 | val_loss=3.4139 val_acc=0.2598 | lr=3.82e-05

Epoch 25/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2567 train_acc=0.2873 | val_loss=3.4035 val_acc=0.2608 | lr=2.68e-05

Epoch 26/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2572 train_acc=0.2882 | val_loss=3.4019 val_acc=0.2608 | lr=1.73e-05

Epoch 27/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2498 train_acc=0.3098 | val_loss=3.4001 val_acc=0.2627 | lr=9.79e-06

Epoch 28/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2341 train_acc=0.3157 | val_loss=3.3956 val_acc=0.2696 | lr=4.37e-06

Epoch 29/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2316 train_acc=0.3127 | val_loss=3.3950 val_acc=0.2676 | lr=1.10e-06

Epoch 30/30


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2346 train_acc=0.2961 | val_loss=3.3948 val_acc=0.2657 | lr=0.00e+00


test:   0%|          | 0/97 [00:00<?, ?it/s]

{'attention': 'standard',
 'params': 745830,
 'best_val_acc': 0.2696078431372549,
 'test_loss': 3.5189838992190334,
 'test_acc': 0.21076597820783868,
 'total_train_time_sec': 731.0605065822601,
 'avg_epoch_train_time_sec': np.float64(16.770416458447773),
 'peak_memory_mb': 10143.72998046875}

Как мы видим, оригинальное внимание при таком числе эпох показывает лучший результат (0.21 test acc против 0.16 на MHLS). При этом MHLS выигрывает по времени (10с на эпоху против 16 у классического внимания) и очень сильно выигрывает по использованию памяти (2610мб против 10143мб у классики).

Как мне кажется, пока эпох слишком мало для сравнения, модели могут еще обучаться. Поставим 60 эпох и посмотрим, что будет.

In [93]:
mhla_model, mhla_history, mhla_summary = run_experiment(
    attention_type="mhla",
    cfg=cfg,
)

mhla_summary



===== Experiment: mhla =====
Params: 751218

Epoch 1/60


/tmp/ipykernel_7000/1961669753.py:35: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = GradScaler(enabled=(cfg.use_amp and device.type == "cuda"))


train:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_7000/1366650589.py:33: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


val:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_7000/1366650589.py:85: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


train_loss=4.6014 train_acc=0.0176 | val_loss=4.4626 val_acc=0.0382 | lr=4.00e-04

Epoch 2/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.4032 train_acc=0.0402 | val_loss=4.3381 val_acc=0.0500 | lr=3.99e-04

Epoch 3/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.2964 train_acc=0.0480 | val_loss=4.2679 val_acc=0.0696 | lr=3.98e-04

Epoch 4/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.2041 train_acc=0.0618 | val_loss=4.1936 val_acc=0.0706 | lr=3.96e-04

Epoch 5/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.1364 train_acc=0.0735 | val_loss=4.1381 val_acc=0.0833 | lr=3.93e-04

Epoch 6/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0678 train_acc=0.0951 | val_loss=4.0731 val_acc=0.0951 | lr=3.90e-04

Epoch 7/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0125 train_acc=0.0980 | val_loss=4.0205 val_acc=0.1118 | lr=3.87e-04

Epoch 8/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.9527 train_acc=0.1049 | val_loss=3.9647 val_acc=0.1255 | lr=3.83e-04

Epoch 9/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.9084 train_acc=0.1167 | val_loss=3.9193 val_acc=0.1157 | lr=3.78e-04

Epoch 10/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8529 train_acc=0.1196 | val_loss=3.8650 val_acc=0.1186 | lr=3.73e-04

Epoch 11/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7813 train_acc=0.1510 | val_loss=3.8322 val_acc=0.1294 | lr=3.68e-04

Epoch 12/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7498 train_acc=0.1422 | val_loss=3.7951 val_acc=0.1353 | lr=3.62e-04

Epoch 13/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6788 train_acc=0.1461 | val_loss=3.7423 val_acc=0.1627 | lr=3.55e-04

Epoch 14/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6548 train_acc=0.1667 | val_loss=3.6997 val_acc=0.1608 | lr=3.49e-04

Epoch 15/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5851 train_acc=0.1755 | val_loss=3.6564 val_acc=0.1725 | lr=3.41e-04

Epoch 16/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5591 train_acc=0.1863 | val_loss=3.6432 val_acc=0.1578 | lr=3.34e-04

Epoch 17/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5156 train_acc=0.1931 | val_loss=3.6205 val_acc=0.1725 | lr=3.26e-04

Epoch 18/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4898 train_acc=0.2049 | val_loss=3.5852 val_acc=0.1912 | lr=3.18e-04

Epoch 19/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4488 train_acc=0.2196 | val_loss=3.5476 val_acc=0.2020 | lr=3.09e-04

Epoch 20/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4049 train_acc=0.2216 | val_loss=3.5216 val_acc=0.2098 | lr=3.00e-04

Epoch 21/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3723 train_acc=0.2343 | val_loss=3.5053 val_acc=0.2069 | lr=2.91e-04

Epoch 22/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3337 train_acc=0.2324 | val_loss=3.4868 val_acc=0.2186 | lr=2.81e-04

Epoch 23/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3171 train_acc=0.2608 | val_loss=3.4922 val_acc=0.1912 | lr=2.72e-04

Epoch 24/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2566 train_acc=0.2735 | val_loss=3.4177 val_acc=0.2422 | lr=2.62e-04

Epoch 25/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2137 train_acc=0.2735 | val_loss=3.4149 val_acc=0.2265 | lr=2.52e-04

Epoch 26/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1952 train_acc=0.2912 | val_loss=3.3841 val_acc=0.2480 | lr=2.42e-04

Epoch 27/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1639 train_acc=0.2971 | val_loss=3.4082 val_acc=0.2373 | lr=2.31e-04

Epoch 28/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1511 train_acc=0.2892 | val_loss=3.3616 val_acc=0.2539 | lr=2.21e-04

Epoch 29/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0969 train_acc=0.3167 | val_loss=3.3454 val_acc=0.2451 | lr=2.10e-04

Epoch 30/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0772 train_acc=0.3020 | val_loss=3.3098 val_acc=0.2657 | lr=2.00e-04

Epoch 31/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0307 train_acc=0.3304 | val_loss=3.3092 val_acc=0.2647 | lr=1.90e-04

Epoch 32/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0008 train_acc=0.3343 | val_loss=3.3050 val_acc=0.2578 | lr=1.79e-04

Epoch 33/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0021 train_acc=0.3314 | val_loss=3.3073 val_acc=0.2559 | lr=1.69e-04

Epoch 34/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9645 train_acc=0.3539 | val_loss=3.2858 val_acc=0.2706 | lr=1.58e-04

Epoch 35/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9541 train_acc=0.3402 | val_loss=3.2725 val_acc=0.2794 | lr=1.48e-04

Epoch 36/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9362 train_acc=0.3373 | val_loss=3.2621 val_acc=0.2824 | lr=1.38e-04

Epoch 37/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8871 train_acc=0.3775 | val_loss=3.2475 val_acc=0.2745 | lr=1.28e-04

Epoch 38/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8819 train_acc=0.3676 | val_loss=3.2488 val_acc=0.2755 | lr=1.19e-04

Epoch 39/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8750 train_acc=0.3520 | val_loss=3.2340 val_acc=0.2824 | lr=1.09e-04

Epoch 40/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8649 train_acc=0.3647 | val_loss=3.2560 val_acc=0.2667 | lr=1.00e-04

Epoch 41/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8425 train_acc=0.3833 | val_loss=3.2326 val_acc=0.2931 | lr=9.11e-05

Epoch 42/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8337 train_acc=0.3784 | val_loss=3.2064 val_acc=0.2990 | lr=8.24e-05

Epoch 43/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8083 train_acc=0.3824 | val_loss=3.2111 val_acc=0.2931 | lr=7.41e-05

Epoch 44/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7988 train_acc=0.3882 | val_loss=3.2018 val_acc=0.2912 | lr=6.62e-05

Epoch 45/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7971 train_acc=0.3941 | val_loss=3.2045 val_acc=0.3000 | lr=5.86e-05

Epoch 46/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7678 train_acc=0.3951 | val_loss=3.1999 val_acc=0.2931 | lr=5.14e-05

Epoch 47/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7522 train_acc=0.4098 | val_loss=3.2009 val_acc=0.2863 | lr=4.46e-05

Epoch 48/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7352 train_acc=0.4029 | val_loss=3.1857 val_acc=0.2980 | lr=3.82e-05

Epoch 49/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7374 train_acc=0.4118 | val_loss=3.1830 val_acc=0.3059 | lr=3.23e-05

Epoch 50/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7376 train_acc=0.4069 | val_loss=3.1781 val_acc=0.3000 | lr=2.68e-05

Epoch 51/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7238 train_acc=0.4167 | val_loss=3.1828 val_acc=0.2902 | lr=2.18e-05

Epoch 52/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7361 train_acc=0.4127 | val_loss=3.1869 val_acc=0.2902 | lr=1.73e-05

Epoch 53/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6960 train_acc=0.4206 | val_loss=3.1829 val_acc=0.2980 | lr=1.33e-05

Epoch 54/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7016 train_acc=0.4304 | val_loss=3.1806 val_acc=0.2980 | lr=9.79e-06

Epoch 55/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7203 train_acc=0.4225 | val_loss=3.1775 val_acc=0.2980 | lr=6.81e-06

Epoch 56/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7057 train_acc=0.4078 | val_loss=3.1770 val_acc=0.2951 | lr=4.37e-06

Epoch 57/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6992 train_acc=0.4147 | val_loss=3.1756 val_acc=0.2961 | lr=2.46e-06

Epoch 58/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7019 train_acc=0.4304 | val_loss=3.1764 val_acc=0.2951 | lr=1.10e-06

Epoch 59/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7156 train_acc=0.4167 | val_loss=3.1757 val_acc=0.2951 | lr=2.74e-07

Epoch 60/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7091 train_acc=0.4216 | val_loss=3.1757 val_acc=0.2951 | lr=0.00e+00


test:   0%|          | 0/97 [00:00<?, ?it/s]

{'attention': 'mhla',
 'params': 751218,
 'best_val_acc': 0.3058823529411765,
 'test_loss': 3.3475339109867757,
 'test_acc': 0.2538624166531143,
 'total_train_time_sec': 962.9993426799774,
 'avg_epoch_train_time_sec': np.float64(10.44358323017756),
 'peak_memory_mb': 2610.66650390625}

In [94]:
standard_model, standard_history, standard_summary = run_experiment(
    attention_type="standard",
    cfg=cfg,
)

standard_summary



===== Experiment: standard =====
Params: 745830

Epoch 1/60


/tmp/ipykernel_7000/1961669753.py:35: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = GradScaler(enabled=(cfg.use_amp and device.type == "cuda"))


train:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_7000/1366650589.py:33: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


val:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_7000/1366650589.py:85: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


train_loss=4.6036 train_acc=0.0137 | val_loss=4.4629 val_acc=0.0431 | lr=4.00e-04

Epoch 2/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.4087 train_acc=0.0382 | val_loss=4.3274 val_acc=0.0598 | lr=3.99e-04

Epoch 3/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.2782 train_acc=0.0569 | val_loss=4.2331 val_acc=0.0843 | lr=3.98e-04

Epoch 4/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.1814 train_acc=0.0696 | val_loss=4.1309 val_acc=0.0892 | lr=3.96e-04

Epoch 5/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0907 train_acc=0.0931 | val_loss=4.0584 val_acc=0.1167 | lr=3.93e-04

Epoch 6/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0235 train_acc=0.1118 | val_loss=3.9818 val_acc=0.1265 | lr=3.90e-04

Epoch 7/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.9428 train_acc=0.1333 | val_loss=3.9170 val_acc=0.1441 | lr=3.87e-04

Epoch 8/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8782 train_acc=0.1284 | val_loss=3.8580 val_acc=0.1686 | lr=3.83e-04

Epoch 9/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7945 train_acc=0.1696 | val_loss=3.7735 val_acc=0.1755 | lr=3.78e-04

Epoch 10/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7325 train_acc=0.1833 | val_loss=3.7257 val_acc=0.1863 | lr=3.73e-04

Epoch 11/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6545 train_acc=0.2010 | val_loss=3.6977 val_acc=0.1765 | lr=3.68e-04

Epoch 12/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6272 train_acc=0.1833 | val_loss=3.6256 val_acc=0.1971 | lr=3.62e-04

Epoch 13/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5582 train_acc=0.2069 | val_loss=3.5878 val_acc=0.2127 | lr=3.55e-04

Epoch 14/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5067 train_acc=0.2225 | val_loss=3.5728 val_acc=0.2010 | lr=3.49e-04

Epoch 15/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4694 train_acc=0.2167 | val_loss=3.5269 val_acc=0.2127 | lr=3.41e-04

Epoch 16/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4280 train_acc=0.2255 | val_loss=3.5461 val_acc=0.2127 | lr=3.34e-04

Epoch 17/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3941 train_acc=0.2343 | val_loss=3.4521 val_acc=0.2294 | lr=3.26e-04

Epoch 18/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3431 train_acc=0.2539 | val_loss=3.4354 val_acc=0.2245 | lr=3.18e-04

Epoch 19/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2961 train_acc=0.2706 | val_loss=3.4467 val_acc=0.2176 | lr=3.09e-04

Epoch 20/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2705 train_acc=0.2598 | val_loss=3.3989 val_acc=0.2412 | lr=3.00e-04

Epoch 21/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2194 train_acc=0.2833 | val_loss=3.3605 val_acc=0.2441 | lr=2.91e-04

Epoch 22/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1622 train_acc=0.3000 | val_loss=3.3421 val_acc=0.2657 | lr=2.81e-04

Epoch 23/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1581 train_acc=0.3029 | val_loss=3.3127 val_acc=0.2657 | lr=2.72e-04

Epoch 24/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1028 train_acc=0.3108 | val_loss=3.2874 val_acc=0.2784 | lr=2.62e-04

Epoch 25/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0654 train_acc=0.3314 | val_loss=3.2531 val_acc=0.2804 | lr=2.52e-04

Epoch 26/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0313 train_acc=0.3294 | val_loss=3.2700 val_acc=0.2657 | lr=2.42e-04

Epoch 27/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0274 train_acc=0.3343 | val_loss=3.2729 val_acc=0.2588 | lr=2.31e-04

Epoch 28/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9559 train_acc=0.3676 | val_loss=3.2094 val_acc=0.3118 | lr=2.21e-04

Epoch 29/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9392 train_acc=0.3843 | val_loss=3.1978 val_acc=0.3069 | lr=2.10e-04

Epoch 30/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8997 train_acc=0.3755 | val_loss=3.1681 val_acc=0.3098 | lr=2.00e-04

Epoch 31/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8661 train_acc=0.3863 | val_loss=3.1473 val_acc=0.3088 | lr=1.90e-04

Epoch 32/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8246 train_acc=0.4157 | val_loss=3.1460 val_acc=0.3235 | lr=1.79e-04

Epoch 33/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8294 train_acc=0.3922 | val_loss=3.1550 val_acc=0.3088 | lr=1.69e-04

Epoch 34/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8293 train_acc=0.3941 | val_loss=3.1352 val_acc=0.3225 | lr=1.58e-04

Epoch 35/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7496 train_acc=0.4176 | val_loss=3.1218 val_acc=0.3206 | lr=1.48e-04

Epoch 36/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7229 train_acc=0.4353 | val_loss=3.1042 val_acc=0.3235 | lr=1.38e-04

Epoch 37/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7216 train_acc=0.4353 | val_loss=3.0889 val_acc=0.3245 | lr=1.28e-04

Epoch 38/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7203 train_acc=0.4402 | val_loss=3.0859 val_acc=0.3245 | lr=1.19e-04

Epoch 39/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6708 train_acc=0.4471 | val_loss=3.0656 val_acc=0.3304 | lr=1.09e-04

Epoch 40/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6736 train_acc=0.4608 | val_loss=3.0625 val_acc=0.3373 | lr=1.00e-04

Epoch 41/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6419 train_acc=0.4578 | val_loss=3.0666 val_acc=0.3304 | lr=9.11e-05

Epoch 42/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6396 train_acc=0.4637 | val_loss=3.0456 val_acc=0.3314 | lr=8.24e-05

Epoch 43/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6045 train_acc=0.4716 | val_loss=3.0336 val_acc=0.3382 | lr=7.41e-05

Epoch 44/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5811 train_acc=0.4735 | val_loss=3.0328 val_acc=0.3353 | lr=6.62e-05

Epoch 45/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5692 train_acc=0.4892 | val_loss=3.0457 val_acc=0.3314 | lr=5.86e-05

Epoch 46/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5681 train_acc=0.4873 | val_loss=3.0275 val_acc=0.3373 | lr=5.14e-05

Epoch 47/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5555 train_acc=0.4912 | val_loss=3.0130 val_acc=0.3422 | lr=4.46e-05

Epoch 48/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5503 train_acc=0.4814 | val_loss=3.0190 val_acc=0.3441 | lr=3.82e-05

Epoch 49/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5288 train_acc=0.5069 | val_loss=2.9990 val_acc=0.3510 | lr=3.23e-05

Epoch 50/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5004 train_acc=0.5176 | val_loss=3.0007 val_acc=0.3500 | lr=2.68e-05

Epoch 51/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5121 train_acc=0.4902 | val_loss=2.9951 val_acc=0.3510 | lr=2.18e-05

Epoch 52/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5078 train_acc=0.5029 | val_loss=2.9915 val_acc=0.3510 | lr=1.73e-05

Epoch 53/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4990 train_acc=0.5167 | val_loss=2.9917 val_acc=0.3461 | lr=1.33e-05

Epoch 54/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5019 train_acc=0.5127 | val_loss=2.9942 val_acc=0.3529 | lr=9.79e-06

Epoch 55/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4932 train_acc=0.5147 | val_loss=2.9960 val_acc=0.3500 | lr=6.81e-06

Epoch 56/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4906 train_acc=0.5059 | val_loss=2.9949 val_acc=0.3480 | lr=4.37e-06

Epoch 57/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4911 train_acc=0.5216 | val_loss=2.9938 val_acc=0.3500 | lr=2.46e-06

Epoch 58/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4767 train_acc=0.5137 | val_loss=2.9928 val_acc=0.3500 | lr=1.10e-06

Epoch 59/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4850 train_acc=0.5010 | val_loss=2.9927 val_acc=0.3490 | lr=2.74e-07

Epoch 60/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4769 train_acc=0.5098 | val_loss=2.9926 val_acc=0.3490 | lr=0.00e+00


test:   0%|          | 0/97 [00:00<?, ?it/s]

{'attention': 'standard',
 'params': 745830,
 'best_val_acc': 0.35294117647058826,
 'test_loss': 3.141749489383012,
 'test_acc': 0.31419743047650023,
 'total_train_time_sec': 1477.7970967292786,
 'avg_epoch_train_time_sec': np.float64(16.96459967692693),
 'peak_memory_mb': 10109.04541015625}

Как мы видим, результаты в плане сравнения примерно такие же. Видно, что обе модели из-за малого числа параметров не могут хорошо обучиться. При этом модель на классическом внимании опять показала лучшую метрику в целом, но сильно проигрывает по времени на эпоху и использованию памяти.
Можно сделать вывод, что внимание MHLS выгодно использовать при малом числе ресурсов (особенно видеопамяти), при этом немного проигрывая в качестве.
Модель с MHLS вниманием можно сделать сильно больше по параметрам, чем модель с классическим вниманием, при этом получить одинаковое использование ресурсов моделями, тем самым возможно получить сопоставимое или лучшее качество.


### ДЗ7

Теперь скачаем репозиторий с SequenceLab вниманием. Эксперименты будем проводить с тем же корпусом, с тем же конфигом, чтобы сравнение было честным.

In [7]:
!git clone https://github.com/LJC-FVNR/SequenceLab.git
%cd SequenceLab
!pip install -e .
%cd ..


Cloning into 'SequenceLab'...
remote: Enumerating objects: 34, done.
remote: Counting objects: 100% (34/34), done.
remote: Compressing objects: 100% (28/28), done.
remote: Total 34 (delta 11), reused 25 (delta 5), pack-reused 0 (from 0)
Receiving objects: 100% (34/34), 30.69 KiB | 10.23 MiB/s, done.
Resolving deltas: 100% (11/11), done.
/content/SequenceLab
Obtaining file:///content/SequenceLab
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Building editable for sequencelab (pyproject.toml) ... done
  Created wheel for sequencelab: filename=sequencelab-0.1.0-0.editable-py3-none-any.whl size=5993 sha256=732b6555ce35bb9757e131ce7c4fd56c80eb8af916f3225f1c0b28c52f5eb5d0
  Stored in directory: /tmp/pip-ephem-wheel-cache-4d0bjx9i/wheels/16/52/55/829630f3803e253dadbec9bc0614b68ab2d7c167cd7e3738ee
Successfully built sequencelab
/cont

In [10]:
import sys
print(sys.executable)


/usr/bin/python3


In [11]:
%pip install -e /content/SequenceLab


Obtaining file:///content/SequenceLab
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Building editable for sequencelab (pyproject.toml) ... done
  Created wheel for sequencelab: filename=sequencelab-0.1.0-0.editable-py3-none-any.whl size=5993 sha256=6e6bd1b0e61787a87c7cda352b0077aec45d714f727cea9a914e8464e1c85d91
  Stored in directory: /tmp/pip-ephem-wheel-cache-484yd6c0/wheels/16/52/55/829630f3803e253dadbec9bc0614b68ab2d7c167cd7e3738ee
Successfully built sequencelab
  Attempting uninstall: sequencelab
    Found existing installation: sequencelab 0.1.0
    Uninstalling sequencelab-0.1.0:
      Successfully uninstalled sequencelab-0.1.0


In [14]:
import sys
import os
import importlib

seq_src = "/content/SequenceLab/src"

if seq_src not in sys.path:
    sys.path.insert(0, seq_src)

importlib.invalidate_caches()

from sequencelab.build import build_attention
from sequencelab.config import WaveConfig, ZeroSConfig, FEMConfig

print("SequenceLab imported successfully")


SequenceLab imported successfully


In [33]:
seq_fem_model, seq_fem_history, seq_fem_summary = run_experiment(
    attention_type="sequencelab_fem_softmax",
    cfg=cfg,
)

seq_fem_summary



===== Experiment: sequencelab_fem_softmax =====
Params: 696678

Epoch 1/60


/tmp/ipykernel_3864/3352663501.py:44: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = GradScaler(enabled=(cfg.use_amp and device.type == "cuda"))


train:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_3864/1366650589.py:33: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


val:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_3864/1366650589.py:85: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


train_loss=4.6250 train_acc=0.0108 | val_loss=4.4667 val_acc=0.0284 | lr=4.00e-04

Epoch 2/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.4107 train_acc=0.0441 | val_loss=4.3257 val_acc=0.0549 | lr=3.99e-04

Epoch 3/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.2771 train_acc=0.0529 | val_loss=4.2295 val_acc=0.0804 | lr=3.98e-04

Epoch 4/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.1815 train_acc=0.0775 | val_loss=4.1356 val_acc=0.0892 | lr=3.96e-04

Epoch 5/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.1004 train_acc=0.0833 | val_loss=4.0685 val_acc=0.0961 | lr=3.93e-04

Epoch 6/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0163 train_acc=0.0980 | val_loss=3.9940 val_acc=0.1353 | lr=3.90e-04

Epoch 7/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.9424 train_acc=0.1225 | val_loss=3.9120 val_acc=0.1441 | lr=3.87e-04

Epoch 8/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8780 train_acc=0.1304 | val_loss=3.8678 val_acc=0.1402 | lr=3.83e-04

Epoch 9/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8104 train_acc=0.1451 | val_loss=3.7972 val_acc=0.1784 | lr=3.78e-04

Epoch 10/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7459 train_acc=0.1598 | val_loss=3.7472 val_acc=0.1676 | lr=3.73e-04

Epoch 11/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6797 train_acc=0.1706 | val_loss=3.7005 val_acc=0.1784 | lr=3.68e-04

Epoch 12/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6169 train_acc=0.1843 | val_loss=3.6526 val_acc=0.1912 | lr=3.62e-04

Epoch 13/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5715 train_acc=0.1941 | val_loss=3.6047 val_acc=0.2000 | lr=3.55e-04

Epoch 14/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5230 train_acc=0.2216 | val_loss=3.5855 val_acc=0.2343 | lr=3.49e-04

Epoch 15/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4691 train_acc=0.2284 | val_loss=3.5273 val_acc=0.2225 | lr=3.41e-04

Epoch 16/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4015 train_acc=0.2422 | val_loss=3.5150 val_acc=0.2275 | lr=3.34e-04

Epoch 17/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3643 train_acc=0.2569 | val_loss=3.4895 val_acc=0.2196 | lr=3.26e-04

Epoch 18/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3139 train_acc=0.2696 | val_loss=3.4341 val_acc=0.2353 | lr=3.18e-04

Epoch 19/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2695 train_acc=0.2657 | val_loss=3.4207 val_acc=0.2314 | lr=3.09e-04

Epoch 20/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2332 train_acc=0.2912 | val_loss=3.3819 val_acc=0.2412 | lr=3.00e-04

Epoch 21/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1797 train_acc=0.3088 | val_loss=3.4086 val_acc=0.2314 | lr=2.91e-04

Epoch 22/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1412 train_acc=0.3167 | val_loss=3.3216 val_acc=0.2814 | lr=2.81e-04

Epoch 23/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0947 train_acc=0.3333 | val_loss=3.3408 val_acc=0.2549 | lr=2.72e-04

Epoch 24/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1057 train_acc=0.3167 | val_loss=3.3013 val_acc=0.2618 | lr=2.62e-04

Epoch 25/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0365 train_acc=0.3392 | val_loss=3.2804 val_acc=0.2696 | lr=2.52e-04

Epoch 26/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0160 train_acc=0.3618 | val_loss=3.2635 val_acc=0.2824 | lr=2.42e-04

Epoch 27/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9957 train_acc=0.3461 | val_loss=3.2304 val_acc=0.2971 | lr=2.31e-04

Epoch 28/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9493 train_acc=0.3578 | val_loss=3.2086 val_acc=0.3000 | lr=2.21e-04

Epoch 29/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8888 train_acc=0.3775 | val_loss=3.1954 val_acc=0.3059 | lr=2.10e-04

Epoch 30/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8818 train_acc=0.3912 | val_loss=3.1754 val_acc=0.2912 | lr=2.00e-04

Epoch 31/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8352 train_acc=0.4000 | val_loss=3.1584 val_acc=0.3108 | lr=1.90e-04

Epoch 32/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8098 train_acc=0.3990 | val_loss=3.1448 val_acc=0.3108 | lr=1.79e-04

Epoch 33/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7970 train_acc=0.3961 | val_loss=3.1477 val_acc=0.3147 | lr=1.69e-04

Epoch 34/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7366 train_acc=0.4088 | val_loss=3.1222 val_acc=0.3108 | lr=1.58e-04

Epoch 35/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7248 train_acc=0.4373 | val_loss=3.1255 val_acc=0.3000 | lr=1.48e-04

Epoch 36/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6837 train_acc=0.4225 | val_loss=3.0860 val_acc=0.3324 | lr=1.38e-04

Epoch 37/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6872 train_acc=0.4324 | val_loss=3.0979 val_acc=0.3245 | lr=1.28e-04

Epoch 38/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6845 train_acc=0.4490 | val_loss=3.0749 val_acc=0.3343 | lr=1.19e-04

Epoch 39/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6394 train_acc=0.4441 | val_loss=3.0692 val_acc=0.3294 | lr=1.09e-04

Epoch 40/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6104 train_acc=0.4618 | val_loss=3.0849 val_acc=0.3284 | lr=1.00e-04

Epoch 41/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5948 train_acc=0.4667 | val_loss=3.0476 val_acc=0.3314 | lr=9.11e-05

Epoch 42/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5836 train_acc=0.4529 | val_loss=3.0614 val_acc=0.3382 | lr=8.24e-05

Epoch 43/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5634 train_acc=0.4814 | val_loss=3.0557 val_acc=0.3422 | lr=7.41e-05

Epoch 44/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5450 train_acc=0.4735 | val_loss=3.0427 val_acc=0.3431 | lr=6.62e-05

Epoch 45/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5102 train_acc=0.4882 | val_loss=3.0475 val_acc=0.3343 | lr=5.86e-05

Epoch 46/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5410 train_acc=0.4931 | val_loss=3.0289 val_acc=0.3510 | lr=5.14e-05

Epoch 47/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4961 train_acc=0.5000 | val_loss=3.0302 val_acc=0.3480 | lr=4.46e-05

Epoch 48/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4849 train_acc=0.5049 | val_loss=3.0263 val_acc=0.3431 | lr=3.82e-05

Epoch 49/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4987 train_acc=0.4971 | val_loss=3.0203 val_acc=0.3480 | lr=3.23e-05

Epoch 50/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4483 train_acc=0.5167 | val_loss=3.0267 val_acc=0.3500 | lr=2.68e-05

Epoch 51/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4710 train_acc=0.5039 | val_loss=3.0212 val_acc=0.3510 | lr=2.18e-05

Epoch 52/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4667 train_acc=0.5118 | val_loss=3.0110 val_acc=0.3520 | lr=1.73e-05

Epoch 53/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4510 train_acc=0.5245 | val_loss=3.0181 val_acc=0.3539 | lr=1.33e-05

Epoch 54/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4357 train_acc=0.5078 | val_loss=3.0155 val_acc=0.3529 | lr=9.79e-06

Epoch 55/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4359 train_acc=0.5167 | val_loss=3.0124 val_acc=0.3578 | lr=6.81e-06

Epoch 56/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4369 train_acc=0.5275 | val_loss=3.0142 val_acc=0.3559 | lr=4.37e-06

Epoch 57/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4438 train_acc=0.5147 | val_loss=3.0151 val_acc=0.3559 | lr=2.46e-06

Epoch 58/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4528 train_acc=0.5196 | val_loss=3.0143 val_acc=0.3559 | lr=1.10e-06

Epoch 59/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4510 train_acc=0.5157 | val_loss=3.0145 val_acc=0.3539 | lr=2.74e-07

Epoch 60/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4325 train_acc=0.5127 | val_loss=3.0143 val_acc=0.3549 | lr=0.00e+00


test:   0%|          | 0/97 [00:00<?, ?it/s]

{'attention': 'sequencelab_fem_softmax',
 'params': 696678,
 'best_val_acc': 0.35784313725490197,
 'test_loss': 3.1643724975827885,
 'test_acc': 0.3081801919011221,
 'total_train_time_sec': 1274.1675953865051,
 'avg_epoch_train_time_sec': np.float64(14.50312800804774),
 'peak_memory_mb': 2578.87158203125}

Как мы видим, модель с вниманием SequenceLab показала примерно такую же точность на тесте, при этом среднее время на эпоху уменьшилось (16с -> 14с), использование памяти так же уменьшилось (10109мб -> 2578мб). По сравнению с вниманием из MHLS точность модели сильно выросла (0.25 acc -> 0.30 acc), при этом время на эпоху у MHLS тоже меньше (10с против 14с), потребление памяти примерно такое же. Можно сделать вывод, что модели с вниманием из SequenceLab сильно уменьшает потребление памяти, уменьшает время на эпоху, при этом почти не ухудшая точность ( в отличие от MHLS).

### ДЗ 8

Делаем все то же самое, что и для прошлых вариантов внимания

In [34]:
# ELSA attention
!git clone https://github.com/ming053l/ELSA.git
!pip install -e ./ELSA


Cloning into 'ELSA'...
remote: Enumerating objects: 114, done.
remote: Counting objects: 100% (114/114), done.
remote: Compressing objects: 100% (113/113), done.
remote: Total 114 (delta 29), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (114/114), 274.78 KiB | 6.24 MiB/s, done.
Resolving deltas: 100% (29/29), done.
Obtaining file:///content/ELSA
  Preparing metadata (setup.py) ... done
  Running setup.py develop for elsa-attention


In [36]:
!pip install timm

import sys
import importlib

elsa_root = "/content/ELSA/code"

if elsa_root not in sys.path:
    sys.path.insert(0, elsa_root)

# чистим возможные неправильные импорты
for name in list(sys.modules.keys()):
    if name == "elsa" or name.startswith("stable"):
        del sys.modules[name]

importlib.invalidate_caches()

from stable.elsa import ElsaAttention

print("ELSA imported successfully")


ELSA imported successfully


/content/ELSA/code/stable/elsa_swin.py:2146: UserWarning: Overwriting elsa_tiny_window16_256 in registry with stable.elsa_swin.elsa_tiny_window16_256. This is because the name being registered conflicts with an existing name. Please check if this is not expected.
  @register_model
/content/ELSA/code/stable/elsa_swin.py:2172: UserWarning: Overwriting elsa_tiny_window8_256 in registry with stable.elsa_swin.elsa_tiny_window8_256. This is because the name being registered conflicts with an existing name. Please check if this is not expected.
  @register_model
/content/ELSA/code/stable/elsa_swin.py:2194: UserWarning: Overwriting elsa_small_window16_256 in registry with stable.elsa_swin.elsa_small_window16_256. This is because the name being registered conflicts with an existing name. Please check if this is not expected.
  @register_model
/content/ELSA/code/stable/elsa_swin.py:2218: UserWarning: Overwriting elsa_small_window8_256 in registry with stable.elsa_swin.elsa_small_window8_256. Thi

In [44]:
elsa_model = elsa_model.to(device)

images, targets = next(iter(train_loader))
images = images.to(device)

with torch.no_grad():
    logits = elsa_model(images)

print(logits.shape)


/content/ELSA/code/stable/elsa.py:1510: RuntimeWarning: backend 'triton' mismatches runtime mode; using 'triton_fp32'.
  primary_backend = self._primary_backend_for_prepack(x)


torch.Size([64, 102])


/content/ELSA/code/stable/elsa.py:1510: RuntimeWarning: backend 'triton' mismatches runtime mode; using 'triton_fp32'.
  primary_backend = self._primary_backend_for_prepack(x)


In [45]:
elsa_model, elsa_history, elsa_summary = run_experiment(
    attention_type="elsa_auto",
    cfg=cfg,
)

elsa_summary



===== Experiment: elsa_auto =====
Params: 744678

Epoch 1/60


/tmp/ipykernel_3864/1060943201.py:50: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = GradScaler(enabled=(cfg.use_amp and device.type == "cuda"))


train:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_3864/1366650589.py:33: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):
/content/ELSA/code/stable/elsa.py:1510: RuntimeWarning: backend 'triton' mismatches runtime mode; using 'triton_fp32_train'.
  primary_backend = self._primary_backend_for_prepack(x)
/content/ELSA/code/stable/elsa.py:1510: RuntimeWarning: backend 'triton' mismatches runtime mode; using 'triton_fp32_train'.
  primary_backend = self._primary_backend_for_prepack(x)
/tmp/ipykernel_3864/1366650589.py:33: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):


val:   0%|          | 0/16 [00:00<?, ?it/s]

/tmp/ipykernel_3864/1366650589.py:85: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with autocast(enabled=(use_amp and device.type == "cuda")):
/content/ELSA/code/stable/elsa.py:1510: RuntimeWarning: backend 'triton' mismatches runtime mode; using 'triton_fp32'.
  primary_backend = self._primary_backend_for_prepack(x)


train_loss=4.6158 train_acc=0.0176 | val_loss=4.4555 val_acc=0.0490 | lr=4.00e-04

Epoch 2/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.3992 train_acc=0.0441 | val_loss=4.3309 val_acc=0.0588 | lr=3.99e-04

Epoch 3/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.2777 train_acc=0.0549 | val_loss=4.2430 val_acc=0.0765 | lr=3.98e-04

Epoch 4/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.1870 train_acc=0.0706 | val_loss=4.1581 val_acc=0.0843 | lr=3.96e-04

Epoch 5/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.1026 train_acc=0.0833 | val_loss=4.0923 val_acc=0.0961 | lr=3.93e-04

Epoch 6/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=4.0433 train_acc=0.0990 | val_loss=4.0203 val_acc=0.1196 | lr=3.90e-04

Epoch 7/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.9657 train_acc=0.1196 | val_loss=3.9526 val_acc=0.1324 | lr=3.87e-04

Epoch 8/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8926 train_acc=0.1402 | val_loss=3.9028 val_acc=0.1127 | lr=3.83e-04

Epoch 9/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.8331 train_acc=0.1431 | val_loss=3.8317 val_acc=0.1471 | lr=3.78e-04

Epoch 10/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7519 train_acc=0.1539 | val_loss=3.7977 val_acc=0.1471 | lr=3.73e-04

Epoch 11/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.7183 train_acc=0.1569 | val_loss=3.7318 val_acc=0.1794 | lr=3.68e-04

Epoch 12/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6558 train_acc=0.1716 | val_loss=3.6923 val_acc=0.1814 | lr=3.62e-04

Epoch 13/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.6224 train_acc=0.1775 | val_loss=3.6657 val_acc=0.1794 | lr=3.55e-04

Epoch 14/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.5513 train_acc=0.2020 | val_loss=3.6165 val_acc=0.1941 | lr=3.49e-04

Epoch 15/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4960 train_acc=0.2118 | val_loss=3.5897 val_acc=0.1980 | lr=3.41e-04

Epoch 16/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4708 train_acc=0.2049 | val_loss=3.5774 val_acc=0.1941 | lr=3.34e-04

Epoch 17/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.4305 train_acc=0.2294 | val_loss=3.5072 val_acc=0.2098 | lr=3.26e-04

Epoch 18/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3697 train_acc=0.2373 | val_loss=3.4745 val_acc=0.2245 | lr=3.18e-04

Epoch 19/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.3524 train_acc=0.2431 | val_loss=3.4484 val_acc=0.2186 | lr=3.09e-04

Epoch 20/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2933 train_acc=0.2578 | val_loss=3.3957 val_acc=0.2569 | lr=3.00e-04

Epoch 21/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2369 train_acc=0.2696 | val_loss=3.3660 val_acc=0.2569 | lr=2.91e-04

Epoch 22/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.2075 train_acc=0.3088 | val_loss=3.3520 val_acc=0.2696 | lr=2.81e-04

Epoch 23/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1602 train_acc=0.3039 | val_loss=3.3378 val_acc=0.2667 | lr=2.72e-04

Epoch 24/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.1144 train_acc=0.3265 | val_loss=3.3013 val_acc=0.2755 | lr=2.62e-04

Epoch 25/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0628 train_acc=0.3422 | val_loss=3.2811 val_acc=0.2833 | lr=2.52e-04

Epoch 26/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0448 train_acc=0.3314 | val_loss=3.2427 val_acc=0.2922 | lr=2.42e-04

Epoch 27/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=3.0042 train_acc=0.3569 | val_loss=3.2380 val_acc=0.2971 | lr=2.31e-04

Epoch 28/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9938 train_acc=0.3471 | val_loss=3.2087 val_acc=0.3000 | lr=2.21e-04

Epoch 29/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9423 train_acc=0.3618 | val_loss=3.2192 val_acc=0.2980 | lr=2.10e-04

Epoch 30/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.9256 train_acc=0.3745 | val_loss=3.1945 val_acc=0.3088 | lr=2.00e-04

Epoch 31/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8730 train_acc=0.3873 | val_loss=3.1857 val_acc=0.3216 | lr=1.90e-04

Epoch 32/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8464 train_acc=0.3941 | val_loss=3.1397 val_acc=0.3343 | lr=1.79e-04

Epoch 33/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8463 train_acc=0.3980 | val_loss=3.1406 val_acc=0.3216 | lr=1.69e-04

Epoch 34/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.8312 train_acc=0.4039 | val_loss=3.1274 val_acc=0.3255 | lr=1.58e-04

Epoch 35/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7717 train_acc=0.4265 | val_loss=3.0994 val_acc=0.3373 | lr=1.48e-04

Epoch 36/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7579 train_acc=0.4284 | val_loss=3.0928 val_acc=0.3275 | lr=1.38e-04

Epoch 37/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7140 train_acc=0.4510 | val_loss=3.0884 val_acc=0.3333 | lr=1.28e-04

Epoch 38/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.7061 train_acc=0.4402 | val_loss=3.0959 val_acc=0.3284 | lr=1.19e-04

Epoch 39/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6692 train_acc=0.4402 | val_loss=3.0759 val_acc=0.3471 | lr=1.09e-04

Epoch 40/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6546 train_acc=0.4735 | val_loss=3.0707 val_acc=0.3382 | lr=1.00e-04

Epoch 41/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.6238 train_acc=0.4627 | val_loss=3.0633 val_acc=0.3412 | lr=9.11e-05

Epoch 42/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5918 train_acc=0.4706 | val_loss=3.0538 val_acc=0.3422 | lr=8.24e-05

Epoch 43/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5976 train_acc=0.4745 | val_loss=3.0514 val_acc=0.3412 | lr=7.41e-05

Epoch 44/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5819 train_acc=0.4843 | val_loss=3.0498 val_acc=0.3451 | lr=6.62e-05

Epoch 45/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5662 train_acc=0.4980 | val_loss=3.0418 val_acc=0.3490 | lr=5.86e-05

Epoch 46/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5550 train_acc=0.4931 | val_loss=3.0409 val_acc=0.3431 | lr=5.14e-05

Epoch 47/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5312 train_acc=0.4971 | val_loss=3.0187 val_acc=0.3539 | lr=4.46e-05

Epoch 48/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5189 train_acc=0.5088 | val_loss=3.0287 val_acc=0.3461 | lr=3.82e-05

Epoch 49/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5215 train_acc=0.5078 | val_loss=3.0210 val_acc=0.3559 | lr=3.23e-05

Epoch 50/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5049 train_acc=0.5216 | val_loss=3.0189 val_acc=0.3461 | lr=2.68e-05

Epoch 51/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4877 train_acc=0.5157 | val_loss=3.0127 val_acc=0.3588 | lr=2.18e-05

Epoch 52/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4800 train_acc=0.4873 | val_loss=3.0122 val_acc=0.3618 | lr=1.73e-05

Epoch 53/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.5073 train_acc=0.5157 | val_loss=3.0070 val_acc=0.3608 | lr=1.33e-05

Epoch 54/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4980 train_acc=0.5098 | val_loss=3.0098 val_acc=0.3637 | lr=9.79e-06

Epoch 55/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4838 train_acc=0.5206 | val_loss=3.0059 val_acc=0.3598 | lr=6.81e-06

Epoch 56/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4731 train_acc=0.5225 | val_loss=3.0049 val_acc=0.3627 | lr=4.37e-06

Epoch 57/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4730 train_acc=0.5206 | val_loss=3.0039 val_acc=0.3598 | lr=2.46e-06

Epoch 58/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4679 train_acc=0.5275 | val_loss=3.0046 val_acc=0.3608 | lr=1.10e-06

Epoch 59/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4804 train_acc=0.5294 | val_loss=3.0049 val_acc=0.3598 | lr=2.74e-07

Epoch 60/60


train:   0%|          | 0/16 [00:00<?, ?it/s]

val:   0%|          | 0/16 [00:00<?, ?it/s]

train_loss=2.4739 train_acc=0.5275 | val_loss=3.0049 val_acc=0.3598 | lr=0.00e+00


test:   0%|          | 0/97 [00:00<?, ?it/s]

{'attention': 'elsa_auto',
 'params': 744678,
 'best_val_acc': 0.3637254901960784,
 'test_loss': 3.1730752886941054,
 'test_acc': 0.3016750691169296,
 'total_train_time_sec': 1240.968228340149,
 'avg_epoch_train_time_sec': np.float64(13.749565776189169),
 'peak_memory_mb': 2059.6767578125}

Модель с вниманием из ELSA показало схожие результаты с моделью с SequenceLab вниманием. Очень похожие результаты по accuracy (в районе 0.3), и времени на эпоху (около 14 секуннд), при этом использование памяти оказалось меньшим (2059мб против 2578мб).

### Итого

| Attention type | Params | Best val acc | Test loss | Test acc | Total train time, sec | Avg epoch train time, sec | Peak memory, MB |
|---|---:|---:|---:|---:|---:|---:|---:|
| MHLA | 751,218 | 0.3059 | 3.3475 | 0.2539 | 963.00 | 10.44 | 2610.67 |
| Standard | 745,830 | 0.3529 | 3.1417 | 0.3142 | 1477.80 | 16.96 | 10109.05 |
| SequenceLab FEM Softmax | 696,678 | 0.3578 | 3.1644 | 0.3082 | 1274.17 | 14.50 | 2578.87 |
| ELSA Auto | 744,678 | 0.3637 | 3.1731 | 0.3017 | 1240.97 | 13.75 | 2059.68 |
|


Лучшее качество на тестовой выборке показала модель с классическим вниманием:

- `test_acc = 0.3142`;
- `best_val_acc = 0.3529`.

Однако классическое внимание оказалось самым дорогим по ресурсам:

- среднее время эпохи — `16.96` секунд;
- пиковое потребление памяти — `10109.05` MB.

Самым быстрым вариантом оказалось MHLA внимание:

- среднее время эпохи — `10.44` секунд;
- общее время обучения — `963.00` секунд.

Но при этом MHLA показало худшее качество:

- `test_acc = 0.2539`;
- `best_val_acc = 0.3059`.

То есть MHLA действительно даёт выигрыш по скорости и памяти, но в данном эксперименте заметно проигрывает по точности классификации.


SequenceLab FEM Softmax показало хороший компромисс между качеством и ресурсами. Его точность почти совпала с классическим attention:

- `test_acc = 0.3082` против `0.3142` у standard attention.

При этом потребление памяти оказалось почти в 4 раза меньше:

- `2578.87` MB против `10109.05` MB.

Среднее время эпохи также меньше, чем у классического attention:

- `14.50` секунд против `16.96` секунд.

ELSA Auto показало лучший результат на validation среди всех моделей:

- `best_val_acc = 0.3637`.

На тесте ELSA немного уступила классическому attention и SequenceLab FEM:

- `test_acc = 0.3017`.

Зато ELSA оказалась самой экономной по памяти:

- `2059.68` MB.

Также ELSA обучалась быстрее стандартного внимания:

- `13.75` секунд на эпоху против `16.96` секунд.

Таким образом, ELSA дала самый сильный выигрыш по памяти при сохранении приемлемого качества. По сравнению с классическим attention она уменьшила пиковое потребление памяти примерно с `10109` MB до `2060` MB, то есть почти в 5 раз, а время эпохи уменьшилось примерно на 19%. При этом падение test accuracy составило около `0.0125`.